File: LRU.py
class Node:
    def __init__(self, key=0, val=0):
        self.key = key
        self.val = val
        self.pre = None
        self.next = None


class DoubleList:
    def __init__(self):
        self.head = Node()
        self.tail = Node()

        self.head.next = self.tail
        self.tail.pre = self.head
        self.size = 0

    # 在链表头部添加x
    def addFirst(self, x):
        tmp_node = self.head.next

        self.head.next = x
        x.pre = self.head
        x.next = tmp_node
        tmp_node.pre = x
        self.size += 1

    # 删除x(x一定是存在的)
    def remove(self, x):
        x.pre.next = x.next
        x.next.pre = x.pre

        x.pre = None
        x.next = None
        self.size -= 1

    # 删除链表中最后一个节点, 并返回该节点
    def remove_last(self):
        x = self.tail.pre
        self.remove(x)
        return x.key, x.val

    def show(self):
        x = self.head.next
        for i in range(self.size):
            print("key: {}, val: {}\n".format(x.key, x.val))
            x = x.next


def dl_usage():
    dl = DoubleList()

    x = Node(3, 4)
    dl.addFirst(x)
    print("add a Node:\n")
    dl.show()

    print("add another Node:\n")
    dl.addFirst(Node(1, 2))
    dl.show()

    dl.remove(x)
    print("remove (3, 4)\n")
    dl.show()

    dl.addFirst(Node(5, 6))
    print("add (5, 6)\n")
    dl.show()

    _, _ = dl.remove_last()
    print("remove last\n")
    dl.show()


class LRUCache:
    def __init__(self, cap):
        self.cap = cap
        self.hash_map = {}
        self.cache = DoubleList()

    def put(self, key, val):
        x = Node(key, val)
        if key in self.hash_map:
            self.cache.remove(self.hash_map.get(key))
            self.cache.addFirst(x)
            self.hash_map[key] = x
        else:
            if self.cap == self.cache.size:
                last = self.cache.tail.pre
                self.hash_map.pop(last.key)
                self.cache.remove_last()
            self.cache.addFirst(x)
            self.hash_map[key] = x

    def get(self, key):
        if key in self.hash_map:
            return -1
        x = self.hash_map[key]
        self.put(x.key, x.val)


def lru_usage():
    lru = LRUCache(3)
    lru.put(1, 2)
    lru.put(3, 4)
    lru.put(1, 3)

    print(lru.hash_map)
    lru.cache.show()
    print("-------------")

    lru.put(5, 6)
    lru.put(7, 8)
    print(lru.hash_map)
    lru.cache.show()


if __name__ == "__main__":
    lru_usage()

File: chapter02/cpp/LRU.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <stdlib.h>
#include <unordered_map>

using namespace std;

struct ListNode
{
    int key;
    int val;
    ListNode* pre;
    ListNode* suc;

    ListNode(int _key=0, int _val=0, ListNode* _pre=nullptr, ListNode* _suc=nullptr) : key(_key), val(_val), pre(_pre), suc(_suc) {}
};

class DoubleList
{
    private:
        int dl_size;
        ListNode* head;
        ListNode* tail;
    
    public:
        DoubleList()
        {
            head = new ListNode();
            tail = new ListNode(0, 0, head);
            head->suc = tail;
            dl_size = 0;
        }

        void addFirst(ListNode* x)
        {
            ListNode* tmp = head->suc;
            head->suc = x;
            x->pre = head;

            x->suc = tmp;
            tmp->pre = x;
            dl_size++;
        }

        void remove(ListNode* x)
        {
            x->pre->suc = x->suc;
            x->suc->pre = x->pre;

            delete x;
            dl_size--;
        }

        vector<int> remove_last()
        {   
            if(dl_size > 0)
            {
                vector<int> v(2, 0);
                v[0] = tail->pre->key;
                v[1] = tail->pre->val;

                remove(tail->pre);
                return v;
            }
            
        }

        void show()
        {   
            ListNode* tmp = head->suc;
            while (tmp != tail)
            {
                cout << "key: " << tmp->key << ", val: " << tmp->val << endl;
                tmp = tmp->suc;
            }
            
        }

        ~DoubleList()
        {
            
            while(head->suc != tail)
            {
                remove(head->suc);
                dl_size--;
            }
            delete head;
            delete tail;
        }

};

void dl_usage()
{
    DoubleList dl;

    cout << "add (3, 4):\n";
    ListNode* x = new ListNode(3, 4);
    dl.addFirst(x);
    dl.show();

    cout << "add (1, 2):\n";
    dl.addFirst(new ListNode(1, 2));
    dl.show();

    cout << "add (5, 6):\n";
    dl.addFirst(new ListNode(5, 6));
    dl.show();

    cout << "remove last: \n";
    dl.remove_last();
    dl.show();
}

class LRUCache
{
    private:
        int cap;
        unordered_map<int, ListNode*> map;
        DoubleList dl;
    
    public:
        LRUCache(int _cap) : cap(_cap) {}

        void put(int key, int val)
        {
            ListNode* x = new ListNode(key, val);
            if(map.count(key))
            {
                dl.remove(map[key]);
                dl.addFirst(x);
                map[key] = x;
            }
            else
            {
                if(cap == map.size())
                {
                    int remove_key = dl.remove_last()[0];
                    map.erase(remove_key);
                }
                dl.addFirst(x);
                map[key] = x;
            }
        }

        int get(int key)
        {
            if(!map.count(key))
            {
                cout << "key not in map\n";
                return -1;
            }
                
            ListNode* x = map[key];
            put(x->key, x->val);
            return x->val;
        }

        void show()
        {
            dl.show();
        }

};

void lru_usage()
{
    LRUCache lru(3);
    lru.put(1, 2);
    lru.put(3, 4);
    lru.put(1, 3);
    cout << "lru show:\n";
    lru.show();

    lru.put(5, 6);
    lru.put(7, 8);

    cout << "lru show:\n";
    lru.show();

    cout << "lru show:\n";
    lru.get(5);
    lru.show();
}


int main()
{
    // dl_usage();
    lru_usage();
    return 0;
}